language/compiler/ir-to-bytecode/syntax/src/syntax.rs

```diff
@@ -1048,81 +1087,98 @@ fn parse_statement(tokens: &mut Lexer) -> Result<Statement, ParseError<Loc, anyh
-    if tokens.peek() == Tok::Else {
-        tokens.advance()?;
-        let else_block = parse_block(tokens)?;
-        Ok(Statement::IfElseStatement(IfElse::if_else(
-            cond, if_block, else_block,
-        )))
-    } else {
-        Ok(Statement::IfElseStatement(IfElse::if_block(cond, if_block)))
-    }
-}
-
-// WhileStatement : Statement = {
-//     "while" "(" <cond: Sp<Exp>> ")" <block: Sp<Block>> => { ... }
-// }
-
-fn parse_while_statement(tokens: &mut Lexer) -> Result<Statement, ParseError<Loc, anyhow::Error>> {
-    consume_token(tokens, Tok::While)?;
-    consume_token(tokens, Tok::LParen)?;
-    let cond = parse_exp(tokens)?;
-    consume_token(tokens, Tok::RParen)?;
-    let block = parse_block(tokens)?;
-    Ok(Statement::WhileStatement(While { cond, block }))
-}
-
-// LoopStatement : Statement = {
-//     "loop" <block: Sp<Block>> => { ... }
-// }
-
-fn parse_loop_statement(tokens: &mut Lexer) -> Result<Statement, ParseError<Loc, anyhow::Error>> {
-    consume_token(tokens, Tok::Loop)?;
-    let block = parse_block(tokens)?;
-    Ok(Statement::LoopStatement(Loop { block }))
+fn parse_block_label(tokens: &mut Lexer) -> Result<BlockLabel, ParseError<Loc, anyhow::Error>> {
+    let start = tokens.start_loc();
+    let name = parse_name(tokens)?;
+    let end = tokens.previous_end_loc();
+    Ok(spanned(tokens.file_name(), start, end, BlockLabel_(name)))
 }
 
 // Statements : Vec<Statement> = {
 //     <Statement*>
 // }
 
-fn parse_statements(tokens: &mut Lexer) -> Result<Vec<Statement>, ParseError<Loc, anyhow::Error>> {
-    let mut stmts: Vec<Statement> = vec![];
-    // The Statements non-terminal in the grammar is always followed by a
-    // closing brace, so continue parsing until we find one of those.
+fn parse_statements(
+    tokens: &mut Lexer,
+) -> Result<Vec<(BlockLabel, Block_)>, ParseError<Loc, anyhow::Error>> {
+    let entry_loc = tokens.previous_end_loc();
+    let mut labels = HashSet::new();
+    let mut blocks = vec![];
+    let mut current_label = None;
+    let mut statements = vec![];
     while tokens.peek() != Tok::RBrace {
-        stmts.push(parse_statement(tokens)?);
+        let statement = parse_statement(tokens)?;
+        match statement {
+            Statement::LabelStatement(label) => {
+                if labels.contains(&label) {
+                    return Err(ParseError::InvalidToken {
+                        location: label.loc,
+                        message: format!("re-declaration of block label {}", label.value),
+                    });
+                }
+                labels.insert(label.clone());
+                blocks.push((
+                    match current_label {
+                        Some(label) => label,
+                        None => spanned(
+                            tokens.file_name(),
+                            entry_loc,
+                            entry_loc,
+                            BlockLabel_(Symbol::from("<entry>")),
```

Comment:
Whats this entry business all about? Just require a label ?

Reply:
I went the other way with it: made the label explicitly optional, so blocks of statements could not have a label.

From the Move IR writer's perspective, requiring a label would look like this, I think:

```diff
  foo() {
+ label_0:
    return;
}
```

This seems like it would require a change for pretty much every `.mvir` test that includes a function implementation. And it's also more inconvenient for test writers, despite not providing any benefit. So instead of implicitly treating the first block of statements in a function body as having the label `<entry>`, I made it have none.